File: src/main/java/com/micabytes/ink/Gather.java
package com.micabytes.ink;

import org.jetbrains.annotations.Nullable;

class Gather extends Container {

  Gather(int l, String str, @Nullable Container current) throws InkParseException {
    lineNumber = l;
    type = ContentType.GATHER;
    level = 2;
    String s = str.substring(1).trim();
    while (s.startsWith("- ")) {
      level++;
      s = s.substring(1).trim();
    }
    if (current == null)
      throw new InkParseException("A gather must be nested within another knot, parent or choice/gather structure");
    parent = current.getContainer(level - 1);
    parent.add(this);
    addLine(s);
  }

  private void addLine(String str) {
    String s = str;
    if (s.startsWith("(")) {
      id = s.substring(s.indexOf(Symbol.BRACE_LEFT) + 1, s.indexOf(Symbol.BRACE_RIGHT)).trim();
      id = (parent != null ? parent.id : null) + InkParser.DOT + id;
      s = s.substring(s.indexOf(Symbol.BRACE_RIGHT) + 1).trim();

    }
    //noinspection ResultOfObjectAllocationIgnored
    new Content(lineNumber, s, this);
  }

  public static boolean isGatherHeader(String str) {
    if (str.length() < 2) return false;
    return str.startsWith("- ");
  }

}

File: src/main/kotlin/com/micabytes/ink/Gather.kt
package com.micabytes.ink

import com.micabytes.ink.exception.InkParseException

internal class Gather @Throws(InkParseException::class)
constructor(text: String,
            parent: Container,
            val level: Int,
            lineNumber: Int) : Container(getId(text, parent), text, parent, lineNumber) {

  init {
    var str = text.substring(1).trim({ it <= ' ' })
    while (str[0] == Symbol.DASH)
      str = str.substring(1).trim({ it <= ' ' })
    if (str.startsWith(Symbol.BRACE_LEFT)) {
      str = str.substring(str.indexOf(Symbol.BRACE_RIGHT) + 1).trim({ it <= ' ' })
    }
    if (!str.isEmpty()) {
      if (str.contains(Symbol.DIVERT))
        InkParser.parseDivert(lineNumber, str, this)
      else
        children.add(Content(Content.getId(this), str, this, lineNumber))
    }
  }

  companion object {

    fun getChoiceDepth(line: String): Int {
      val notation = line[0]
      var lvl = 0
      var s = line.substring(1).trim({ it <= ' ' })
      while (s[0] == notation) {
        lvl++
        s = s.substring(1).trim({ it <= ' ' })
      }
      return lvl
    }

    fun getId(text: String, parent: Container): String {
      var str = text.substring(1).trim({ it <= ' ' })
      while (str[0] == Symbol.DASH)
        str = str.substring(1).trim({ it <= ' ' })
      if (str.startsWith(Symbol.BRACE_LEFT)) {
        val id = str.substring(str.indexOf(Symbol.BRACE_LEFT) + 1, str.indexOf(Symbol.BRACE_RIGHT)).trim({ it <= ' ' })
        return parent.id + Symbol.DOT + id
      }
      return parent.id + Symbol.DOT + parent.size
    }

    fun getParent(currentContainer: Container, lvl: Int): Container {
      var ret = currentContainer
      val parentFound = false
      while (!parentFound) {
        if ((ret is Choice && lvl <= ret.level) || (ret is Gather && lvl <= ret.level))
          ret = ret.parent!!
        else
          return ret
      }
      return ret
    }
  }


}
